sequitur.go

```diff
@@ -323,21 +337,60 @@ func (g *Grammar) PrettyPrint(w io.Writer) error {
 // ErrAlreadyParsed is returned if the grammar instance has already parsed a grammar
 var ErrAlreadyParsed = errors.New("sequitor: grammar already parsed")
 
-// Parse parses a byte string
-func (g *Grammar) Parse(str []byte) error {
+// ErrEmptyInput is returned if the input string is empty
+var ErrEmptyInput = errors.New("sequitor: empty input")
+
+// ErrMalformedUTF8 is returned if the input string contains malformed UTF-8
+var ErrMalformedUTF8 = errors.New("sequitor: malformed utf-8 input")
+
+// ParseUTF8 parses a byte string using utf-8 encoding
+func ParseUTF8(str []byte) (*Grammar, error) {
+	g := new(Grammar)
+	return g, g.parse(str, true)
+}
+
+// ParseBinary parses a binary byte string
+func ParseBinary(str []byte) (*Grammar, error) {
+	g := new(Grammar)
+	return g, g.parse(str, false)
+}
+
+func (g *Grammar) parse(str []byte, useUTF8 bool) error {
 	if g.base != nil {
 		return ErrAlreadyParsed
 	}
+	if len(str) == 0 {
+		return ErrEmptyInput
+	}
 
-	g.ruleID = 256
+	g.ruleID = uint64(utf8.MaxRune) + 1 // larger than the largest rune
 	g.table = make(digrams)
 	g.base = g.newRules()
 
-	g.base.last().insertAfter(g.newSymbolFromValue(uint64(str[0])))
-
-	for _, c := range str[1:] {
-		g.base.last().insertAfter(g.newSymbolFromValue(uint64(c)))
-		g.base.last().prev.check()
+	off := 0
+	for {
```

Comment:
Rather than having two bounds checks, you could do:

     for off < len(str) {

and avoid the offset checks below (DecodeRune can't return zero on a non-empty input)

Reply:
After the last rune is consumed, DecodeRune is given an empty input the next time round.

Reply:
Not if off < len(str) (as suggested for a loop invariant), surely?

Reply:
You're probably right Roger, a function of my refactoring utf-8/binary dual paths, so another issue for @dgryski to look at.